### ISSUE TITLE
Can I create indexes after initial load?

### ISSUE THREAD

**Issue by leandrotocachelis** [2ndQuadrant/pglogical#344]
Initial load, a lot of contention is taking because of the indexes on target database.

Can I pause subscription after COPY finish?

Ideal steps like this:

- create subscription
- start initial load (COPY)
- disable subscription when COPY finish
- create index on target manually
- enable subscription after create index finish

It's possible?

Or another best ideia to share?

Thanks.

**Comment by leandrotocachelis** [comment]
Another idea I have like this:

create_subscription with new parameter example:

pglogical.create_subscription(subscription_name name, provider_dsn text, replication_sets text[], synchronize_structure boolean, synchronize_data boolean, forward_origins text[], apply_delay interval, **initial_load_manual boolean**)

When `initial_load_manual = true`
`create_subscription` start paused and returns `snapshot_ID` of source database at time run `create_replication_slot`

And now we connect on source database and start COPY with `SET TRANSACTION SNAPSHOT snapshot_ID`

When COPY manual finish, we enable subscription to start logical replication

Thanks.

**Comment by eulerto** [comment]
You have 2 requests:

(a) pause replication after initial load;
(b) don't load data. Instead, let user load it manually.

I don't understand what (a) solves. You can always create indexes **after** the initial load. If you are concerned about index build that locks out writes, you can always use CONCURRENTLY. Your approach does not consider the backlog (for example, if you are COPYing large tables) that should be applied after the COPY and before you create indexes.

What's the point of (b)? There is not such COPY syntax. It seems you are complicating a replication setup with no benefit.

In summary, you can use `pglogical.wait_for_subscription_sync_complete` that waits until the data sync is done. After that, you can disable the subscription, build your indexes and enable it again.

**Comment by leandrotocachelis** [comment]
Thanks @eulerto, your suggestion resolved the problem.

My steps was:

1. pg_dump/pg_restore only pre-data: `pg_dump --section=pre-data -n public -Fc -h HOSTNAME -p 5432 -d DATABASE -U USERNAME | pg_restore -h HOSTNAME -U USERNAME -d DATABASE`
2. Create node in provider and subscriber
3. Create replication set on provider
4. Add table in replication set
5. Create subscription on subscriber with `synchronize_data := true`
6. While run initial load (COPYing large tables), I ran: `select pglogical.wait_for_subscription_sync_complete('SUBSCRIPTION_NAME'); select pglogical.alter_subscription_disable('SUBSCRIPTION_NAME', true);`
7. After step 6 finished, I ran pg_dump/pg_restore only post-data: `pg_dump --section=post-data -n public -Fc -h HOSTNAME -p 5432 -d DATABASE -U USERNAME | pg_restore -h HOSTNAME -U USERNAME -d DATABASE`
8. After step 7 finished, I enabled subscription and the data that was dammed was successfully applied.
9. Now logical replication is delayed close to 0.

Thanks.

**Comment by joshuabaird** [comment]
@eulerto Could you please clarify your suggestion?  I'm also running into slow initial sync issues due to indexes.  If I understand correctly, I think you are suggesting:

* Loading schema on the subscriber
* Dropping indexes on subscriber
* Create subscription with data sync enabled (won't sync indexes as they have been dropped on the subscriber)
* Once sync finishes, rebuild indexes on the subscriber

If this is correct, is it necessary to disable the subscription while the indexes are rebuilt on the subscriber?

**Comment by eulerto** [comment]
No. However, if you don't use `CREATE INDEX CONCURRENTLY` while building an index on a large table, you will observe replication lag because an index build blocks concurrent data changes or even a long running transaction (that modifies data) can block an index build. That's why I suggested to disable the subscription to complete the index building task as soon as possible.

**Comment by joshuabaird** [comment]
@eulerto Got it, thanks.  As a test, I have:

* Loaded `pre-data` on the subscriber
* Created primary keys (copied them from `post-data`) but didn't copy other FK/Indexes
* Created a subscription

The subscription status is "replicating," but no data is being copied.  A `COPY` is not running.  On the provider, we do see a replication slot:

```
-[ RECORD 3 ]-------+-----------------------------------------------
slot_name           | pgl_checkout_provider_subscription
plugin              | pglogical_output
slot_type           | logical
datoid              | 16400
database            | checkout
active              | t
active_pid          | 67828
xmin                |
catalog_xmin        | 1199332928
restart_lsn         | 4D4C/9D633B60
confirmed_flush_lsn |
```

Any idea what may be happening?  Not seeing anything in the logs.  I have debug logs enabled on the subscriber:

```
2021-11-04 15:52:57 UTC:10.178.68.249(42802):platform@checkout:[8016]:LOG:  duration: 2047.497 ms  statement: SELECT pglogical.create_subscription(
	  subscription_name := 'subscription',
	  provider_dsn := 'xyz'
	);
2021-11-04 15:52:57 UTC::@:[30988]:DEBUG:  StartTransaction(1) name: unnamed; blockState: DEFAULT; state: INPROGRESS, xid/subid/cid: 0/1/0
2021-11-04 15:52:57 UTC::@:[30988]:DEBUG:  registering pglogical manager process for database template1
2021-11-04 15:52:57 UTC::@:[30924]:DEBUG:  registering background worker "pglogical manager 1"
2021-11-04 15:52:57 UTC::@:[30924]:DEBUG:  registering background worker "pglogical apply 16397:2875150205"
2021-11-04 15:52:57 UTC::@:[30924]:DEBUG:  starting background worker process "pglogical apply 16397:2875150205"
2021-11-04 15:52:57 UTC::@:[30924]:DEBUG:  starting background worker process "pglogical manager 1"
````

**Comment by joshuabaird** [comment]
It looks like this was related to trying to subscribe two subscribers with a single provider.  Not sure what is up there, but I have a sync now running between a single provider and subscriber without indexes and FK's.  We'll see how it goes!
